homeassistant/components/homekit/type_sensors.py

```diff
@@ -128,21 +153,244 @@ def __init__(self, *args):
         """Initialize a AirQualitySensor accessory object."""
         super().__init__(*args, category=CATEGORY_SENSOR)
 
-        serv_air_quality = self.add_preload_service(
-            SERV_AIR_QUALITY_SENSOR, [CHAR_AIR_PARTICULATE_DENSITY]
+        """determine the charciteristics of the sensor based on the curent attributes"""
+        chars = []
+        current_state = self.hass.states.get(self.entity_id)
+        nitrogen_dioxide_density = convert_to_float(
+            current_state.attributes.get(ATTR_NITROGEN_DIOXIDE_DENSITY)
+        )
+        voc_density = convert_to_float(current_state.attributes.get(ATTR_VOC_DENSITY))
+        pm_2_5_density = convert_to_float(
+            current_state.attributes.get(ATTR_PM_2_5_DENSITY)
         )
-        self.char_quality = serv_air_quality.configure_char(CHAR_AIR_QUALITY, value=0)
-        self.char_density = serv_air_quality.configure_char(
-            CHAR_AIR_PARTICULATE_DENSITY, value=0
+        pm_10_density = convert_to_float(
+            current_state.attributes.get(ATTR_PM_10_DENSITY)
+        )
+        pm_density = convert_to_float(current_state.attributes.get(ATTR_PM_DENSITY))
+        pm_size = convert_to_float(current_state.attributes.get(ATTR_PM_SIZE))
+
+        if nitrogen_dioxide_density is not None:
+            chars.append(CHAR_NITROGEN_DIOXIDE_DENSITY)
+
+        if voc_density is not None:
+            chars.append(CHAR_VOC_DENSITY)
+
+        """
+        NOTE: this would be the easy way but it seems the driver is missing the PM2_5_Density key
+              because trying to do this causes the following when tyring to call self.add_preload_service:
+
+                File "/usr/local/lib/python3.7/site-packages/pyhap/loader.py", line 42, in get_char
+                  char_dict = self.char_types[name].copy()
+              KeyError: 'PM2_5Density'
+
+        if pm_2_5_density is not None:
+            chars.append(CHAR_PM_2_5_DENSITY)
+        """
+
+        if pm_10_density is not None:
+            chars.append(CHAR_PM_10_DENSITY)
+
+        if (
+            (pm_density is not None)
```

Comment:
You don't normally need so many parenthesis in python, you should do:

```
if pm_density is not None and pm_2_5_density is None and pm_10_density is not None:
    chars.append(...)
```

I think this one will actually fit on one line without the extra parenthsis there. There are a few others that look like this applies to too. Can you update them too.

Reply:
all the parenthesis were for my brain, but i re-wrote this section so now its much simpler and preserves the "legacy" code path better.